### PR TITLE
[CLOUDTRUST-2108] Physical Identication API

### PR DESCRIPTION
**Pull request by fperot74**
[CLOUDTRUST-2224] Add missing fields from /account

Uses cloudtrust/common-service#36

### CODE REVIEW

File: api/register/api.go [hunk 1]
```diff
@@ -11,27 +12,28 @@ import (
...
 
+// ActionRepresentation struct
+type ActionRepresentation struct {
+	Name  *string `json:"name"`
+	Scope *string `json:"scope"`
+}
+
 // User representation
 type User struct {
+	UserName             *string `json:"userName,omitempty"`
```
Comment (bsoniam) [inline api/register/api.go]: In some places you have UserName and in others Username (api/account/api.go)

File: api/register/api.go [hunk 2]
```diff
@@ -78,8 +80,8 @@ func (u *User) UserToJSON() string {
 	return string(bytes)
 }
 
-// UpdateUserRepresentation converts a given User to a Keycloak UserRepresentation
-func (u *User) UpdateUserRepresentation(kcUser *kc.UserRepresentation) {
+// ExportToKeycloak converts a given User to a Keycloak UserRepresentation
```
[the remark this answers is not in the document]
Comment (fperot74) [reply api/register/api.go]: It is not exactly the same thing: Convert converts a structure to another one. Here, I import details of a structure in another one. I need to import values from 2 different origins (Keycloak and DB) so I can't simply convert values

File: api/register/api.go [hunk 3]
```diff
@@ -11,27 +12,28 @@ import (
 	"github.com/nyaruka/phonenumbers"
 )
 
+// ActionRepresentation struct
+type ActionRepresentation struct {
```
[the remark this answers is not in the document]
Comment (fperot74) [reply api/register/api.go]: Not yet merged

File: api/register/api.go (hunk 1 above)
Comment (fperot74) [reply api/register/api.go]: Kept Username

File: api/register/api.go [hunk 4]
```diff
@@ -11,27 +12,28 @@ import (
...
-}
-
-// DBUser struct
-type DBUser struct {
-	UserID               *string `json:"-"`
-	BirthLocation        *string `json:"birth_location,omitempty"`
-	IDDocumentType       *string `json:"id_document_typ,omitempty"`
-	IDDocumentNumber     *string `json:"id_document_num,omitempty"`
-	IDDocumentExpiration *string `json:"id_document_exp,omitempty"`
+	Comment              *string `json:"comment,omitempty"`
```
[the remark this answers is not in the document]
Comment (fperot74) [reply api/register/api.go]: register has been splitted into register + kyc

File: internal/keycloakb/usersdbmodule.go [hunk 5]
```diff
@@ -42,7 +42,26 @@ func NewUsersDBModule(db sqltypes.CloudtrustDB, logger log.Logger) UsersDBModule
 	}
 }
 
-func (c *usersDBModule) StoreOrUpdateUser(ctx context.Context, realm string, user apiregister.DBUser) error {
+// AppendValidationInMemory appends in memory the given validation to the existing validations of the given user
+func (c *usersDBModule) AppendValidationInMemory(ctx context.Context, realm string, user *dto.DBUser, validation dto.DBValidation) error {
```
[the remark this answers is not in the document]
Comment (fperot74) [reply internal/keycloakb/usersdbmodule.go]: Validation item is add in memory but added to items we get from database...

File: api/register/api.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (fperot74) [reply api/register/api.go]: I was able to refactor this part and use a ConvertToKeycloak() function here... (but not possible for api/kyc)

File: cmd/keycloakb/keycloak_bridge.go [hunk 6]
```diff
@@ -1142,12 +1184,21 @@ func configureAccountHandler(ComponentName string, ComponentID string, idGenerat
...
+		return handler
+	}
+}
+
 func configureRegisterHandler(ComponentName string, ComponentID string, idGenerator idgenerator.IDGenerator, keycloakClient *keycloak.Client, recaptchaURL, recaptchaSecret string, tracer tracing.OpentracingClient, logger log.Logger) func(endpoint endpoint.Endpoint) http.Handler {
 	return func(endpoint endpoint.Endpoint) http.Handler {
 		var handler http.Handler
 		handler = register.MakeRegisterHandler(endpoint, logger)
 		handler = middleware.MakeHTTPCorrelationIDMW(idGenerator, tracer, logger, ComponentName, ComponentID)(handler)
-		handler = register.MakeHTTPRecaptchaValidationMW(recaptchaURL, recaptchaSecret, logger)(handler)
```
Comment (harture) [inline cmd/keycloakb/keycloak_bridge.go]: Why we remove recaptcha MW ?

File: pkg/kyc/component.go [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+	configDBModule ConfigurationDBModule
+	eventsDBModule database.EventsDBModule
+	logger         internal.Logger
+}
+
+// NewComponent returns the management component.
+func NewComponent(keycloakClient KeycloakClient, tokenProvider keycloak.OidcTokenProvider, usersDBModule keycloakb.UsersDBModule, configDBModule ConfigurationDBModule, eventsDBModule database.EventsDBModule, logger internal.Logger) Component {
+	return &component{
+		keycloakClient: keycloakClient,
+		tokenProvider:  tokenProvider,
```
Comment (harture) [inline pkg/kyc/component.go]: For KYC, I think tokenProvider should not be needeed as the user must has rights to perform the action

File: pkg/register/http.go [hunk 8]
```diff
@@ -10,19 +10,25 @@ import (
 	http_transport "github.com/go-kit/kit/transport/http"
 )
 
+// RegistContext type
+type RegistContext int
+
+// Regular expressions
 const (
-	// RegExpRealmName is a regular expression for realm names
+	RegistContextRecaptcha RegistContext = iota
```
Comment (harture) [inline pkg/register/http.go]: I don't understand what is it and where we use it ?

File: pkg/register/http.go [hunk 9]
```diff
@@ -10,19 +10,25 @@ import (
...
+
 	RegExpRealmName = `^[a-zA-Z0-9_-]{1,36}$`
 )
 
 // MakeRegisterHandler make an HTTP handler for the self-register endpoint.
 func MakeRegisterHandler(e endpoint.Endpoint, logger log.Logger) *http_transport.Server {
-	pathParams := map[string]string{"realm": RegExpRealmName}
+	pathParams := map[string]string{}
 	queryParams := map[string]string{"realm": RegExpRealmName}
+	headers := []string{"Authorization"}
```
Comment (harture) [inline pkg/register/http.go]: Why component would need it ?

File: pkg/register/authorization.go [hunk 10]
```diff
@@ -27,51 +26,58 @@ type recaptchaResponse struct {
 	ErrorCodes  []string `json:"error-codes"`
 }
 
-// MakeHTTPRecaptchaValidationMW retrieves the recaptcha code and checks its validity
-func MakeHTTPRecaptchaValidationMW(recaptchaURL string, recaptchaSecret string, logger log.Logger) func(http.Handler) http.Handler {
```
Comment (harture) [inline pkg/register/authorization.go]: Why we remove this MW ?

File: pkg/register/authorization.go [hunk 11]
```diff
@@ -27,51 +26,58 @@ type recaptchaResponse struct {
...
+			logger:          logger,
+			next:            next,
+		}
 	}
 }
 
-func httpErrorHandler(_ context.Context, statusCode int, err error, w http.ResponseWriter) {
-	w.WriteHeader(statusCode)
-	w.Header().Set("Content-Type", "application/json; charset=utf-8")
+func (c *authorizationComponentMW) RegisterUser(ctx context.Context, socialRealmName, clientRealmName string, user apiregister.UserRepresentation) (string, error) {
```
Comment (harture) [inline pkg/register/authorization.go]: From my point of view, the recaptcha should be check at a higher level. I was more convinced by the middleware mechanism

File: pkg/register/component.go [hunk 12]
```diff
@@ -35,30 +36,25 @@ type ConfigurationDBModule interface {
 
 // Component is the register component interface.
 type Component interface {
-	RegisterUser(ctx context.Context, realmName string, user apiregister.User) (string, error)
-	GetConfiguration(ctx context.Context, realmName string) (apiregister.Configuration, error)
+	RegisterUser(ctx context.Context, socialRealmName, clientRealmName string, user apiregister.UserRepresentation) (string, error)
```
Comment (harture) [inline pkg/register/component.go]: socialRealmName is a fixed value so I think we could put it in the component struct

File: pkg/register/component.go [hunk 13]
```diff
@@ -102,29 +98,39 @@ func (c *component) RegisterUser(ctx context.Context, realmName string, user api
...
-			// Create the user in Keycloak
-			kcUser = &kc.UserRepresentation{}
 			kcUser.Username = &username
-			user.UpdateUserRepresentation(kcUser)
-			(*kcUser.Attributes)["trustIDAuthToken"] = []string{authToken}
 
-			userID, err = c.keycloakClient.CreateUser(accessToken, c.realm, c.realm, *kcUser)
+			userID, err = c.keycloakClient.CreateUser(accessToken, realmName, realmName, kcUser)
+
 			// Create success: just have to get the userID and exist this loop
```
Comment (harture) [inline pkg/register/component.go]: exit

File: internal/keycloakb/usersdbmodule.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (fperot74) [reply internal/keycloakb/usersdbmodule.go]: Already done

File: cmd/keycloakb/keycloak_bridge.go [hunk 14]
```diff
@@ -592,15 +598,39 @@ func main() {
...
 
 		// new module for register service
-		registerComponent := register.NewComponent(keycloakPublicURL, registerRealm, keycloakClient, oidcTokenProvider, usersDBModule, configDBModule, eventsDBModule, registerLogger)
-		registerComponent = register.MakeAuthorizationRegisterComponentMW(log.With(registerLogger, "mw", "endpoint"))(registerComponent)
+		registerComponent := register.NewComponent(keycloakPublicURL, keycloakClient, oidcTokenProvider, usersDBModule, configDBModule, eventsDBModule, registerLogger)
+		registerComponent = register.MakeAuthorizationRegisterComponentMW(recaptchaURL, recaptchaSecret, log.With(registerLogger, "mw", "endpoint"))(registerComponent)
 
 		registerEndpoints = register.Endpoints{
-			RegisterUser:     prepareEndpoint(register.MakeRegisterUserEndpoint(registerComponent), "register_user", influxMetrics, registerLogger, tracer, rateLimit["register"]),
-			GetConfiguration: prepareEndpoint(register.MakeGetConfigurationEndpoint(registerComponent), "get_configuration", influxMetrics, registerLogger, tracer, rateLimit["register"]),
```
[the remark this answers is not in the document]
Comment (fperot74) [reply cmd/keycloakb/keycloak_bridge.go]: Sorry... may be lost during rebase

File: cmd/keycloakb/keycloak_bridge.go [hunk 15]
```diff
@@ -592,15 +598,39 @@ func main() {
 		var configDBModule = createConfigurationDBModule(configurationRwDBConn, influxMetrics, registerLogger)
 
 		// module for storing and retrieving details of the self-registered users
-		var usersDBModule = register.NewUsersDBModule(usersRwDBConn, registerLogger)
+		var usersDBModule = keycloakb.NewUsersDBModule(usersRwDBConn, registerLogger)
 
 		// new module for register service
-		registerComponent := register.NewComponent(keycloakPublicURL, registerRealm, keycloakClient, oidcTokenProvider, usersDBModule, configDBModule, eventsDBModule, registerLogger)
-		registerComponent = register.MakeAuthorizationRegisterComponentMW(log.With(registerLogger, "mw", "endpoint"))(registerComponent)
+		registerComponent := register.NewComponent(keycloakPublicURL, keycloakClient, oidcTokenProvider, usersDBModule, configDBModule, eventsDBModule, registerLogger)
```
[the remark this answers is not in the document]
Comment (fperot74) [reply cmd/keycloakb/keycloak_bridge.go]: It was due to some changes that have now been rolled back

File: api/kyc/swagger-api_kyc.yaml [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+          description: Successful operation. Returns the generated username
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/User'
+        403:
+          description: No permission to call this operation
+        404:
+          description: User not found
+    post:
```
[the remark this answers is not in the document]
Comment (fperot74) [reply api/kyc/swagger-api_kyc.yaml]: Swagger not in phase with code... Thanks

File: cmd/keycloakb/keycloak_bridge.go (hunk 6 above)
Comment (fperot74) [reply cmd/keycloakb/keycloak_bridge.go]: Because it was done at authorization level... rolled back

File: pkg/kyc/component.go (hunk 7 above)
Comment (fperot74) [reply pkg/kyc/component.go]: Forgot to remove that when splitting previous version

File: pkg/register/http.go (hunk 8 above)
Comment (fperot74) [reply pkg/register/http.go]: Removed

File: pkg/register/http.go (hunk 9 above)
Comment (fperot74) [reply pkg/register/http.go]: Rolled back

File: pkg/register/authorization.go (hunk 10 above)
Comment (fperot74) [reply pkg/register/authorization.go]: Rolled back

File: pkg/register/authorization.go (hunk 11 above)
Comment (fperot74) [reply pkg/register/authorization.go]: Rolled back

File: pkg/register/component.go (hunk 12 above)
Comment (fperot74) [reply pkg/register/component.go]: Rolled back

File: cmd/keycloakb/keycloak_bridge.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (harture) [reply cmd/keycloakb/keycloak_bridge.go]: No problem

File: cmd/keycloakb/keycloak_bridge.go (hunk 6 above)
Comment (harture) [reply cmd/keycloakb/keycloak_bridge.go]: thanks

File: pkg/register/http.go (hunk 9 above)
Comment (harture) [reply pkg/register/http.go]: thanks

File: pkg/kyc/authorization.go [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+
+type authorizationComponentMW struct {
+	realmName   string
+	authManager security.AuthorizationManager
+	logger      log.Logger
+	next        Component
+}
+
+// MakeAuthorizationRegisterComponentMW checks authorization and return an error if the action is not allowed.
+func MakeAuthorizationRegisterComponentMW(realmName string, logger log.Logger, authorizationManager security.AuthorizationManager) func(Component) Component {
```
Comment (harture) [inline pkg/kyc/authorization.go]: MakeAuthorizationKYCComponentMW

**Pull request merged**